### DISCUSSION TITLE
`UnitOfWork::getEntityChangeSet` violates contract

### DISCUSSION THREAD

**Discussion by norkunas** [doctrine/doctrine2#8955] [General]
Maybe I don't understand how things should work with collections, but currently this is happening:

`UnitOfWork::getEntityChangeSet` method defines return as:
```
@return mixed[][]
@psalm-return array<string, array{mixed, mixed}>
```

But for example on this entity:
```
class Entity {
  public Collection $attributes;

  public function setAttributes(Collection $attributes) { $this->attributes = $attributes; }
}
```

Calling `setAttributes()` with a new collection then later to populate changeset calling `UnitOfWork::computeChangeSets` and then `UnitOfWork::getEntityChangeSet` returns `array{attributes: Doctrine\ORM\PersistentCollection}`

**Comment by greg0ire** [comment]
I made that change at part of https://github.com/doctrine/orm/pull/8484 , which is one among many, many PRs, and it's very possible that I messed up . Can you please send a PR to fix the situation? :pray:

**Reply by norkunas** [reply]
I'm not sure what should be fixed, as you haven't changed the code but only added @psalm-var, so I guess the typehint is now wrong? Or UoW fails to return the expected structure?

**Reply by greg0ire** [reply]
Yes, the typehint is most probably wrong, since I added very hastily.

**Reply by norkunas** [reply]
Should it be also same in https://github.com/doctrine/orm/blob/2.9.x/lib/Doctrine/ORM/Event/PreUpdateEventArgs.php#L51 ?

**Reply by greg0ire** [reply]
Given the Psalm failure you get here, I would say yes: https://github.com/doctrine/orm/pull/8959/checks?check_run_id=3419339848

**Reply by norkunas** [reply]
I've checked source so this event is instantiated with `$this->getEntityChangeSet($entity)` so imho they should be same

**Reply by greg0ire** [reply]
Ok, then please change it too :pray:
